### PR TITLE
Feature/t2t support and trainable binf

### PR DESCRIPTION
**Pull request by EugKar**
* Removing binary features from `input_fn`, as soon as all our datasets contain phonemes, binary features are obtained via mapping matrix.
* Adding `--t2t_format` flag to use with datasets collected for Tensor2Tensor framework, e.g. [CommonVoice](https://github.com/sciforce/t2t_problems) problem.
* Fixing `build_vocab` for Tensor2Tensor format to enable getting  maximal lengths.
* Fixing parameters save/load at TPU training.
* Enabling trainable binary features via `--binf_trainable` (may have conflicts with other branches).

### CODE REVIEW

File: infer.py [hunk 1]
```diff
@@ -254,7 +227,7 @@ def model_fn(features, labels,
     predictions = model.predict(
         input_fn=lambda: input_fn(
```
Comment (akademi4eg) [inline infer.py]: Where does it take `input_fn` from if it is removed from this file?

File: model_helper.py [hunk 2]
```diff
@@ -335,6 +339,16 @@ def las_model_fn(features,
             output_dir=os.path.join(config.model_dir, run_name),
             summary_op=attention_summary)
         hooks = [eval_summary_hook]
+        if binf_embedding is not None:
+            with tf.name_scope('binf_image'):
+                binf_image = tf.tile(binf_embedding[None, :, :, None], [tf.shape(encoder_inputs)[0], 1, 1, 1])
```
Comment (akademi4eg) [inline model_helper.py]: Why tile it for a whole batch? Isn't `tf.expand_dims(tf.expand_dims(binf_embedding, axis=0), axis=-1)` simpler?

File: train.py [hunk 3]
```diff
@@ -1,5 +1,8 @@
 import argparse
 import tensorflow as tf
+import multiprocessing
+import os
+from tensor2tensor.data_generators.text_encoder import PAD_ID, EOS_ID
```
Comment (akademi4eg) [inline train.py:5]: Maybe adding `tensor2tensor` as unconditioned dependency is too much? Especially if we just use the dataset generated by it?

File: train.py [hunk 4]
```diff
@@ -176,33 +158,39 @@ def model_fn(features, labels,
...
+                num_epochs=args.num_epochs if mode == tf.estimator.ModeKeys.TRAIN else 1,
+                num_parallel_calls=64 if args.tpu_name and args.tpu_name != 'fake' else args.num_parallel_calls,
+                max_frames=args.max_frames, max_symbols=args.max_symbols)
+        else:
+            return lambda params: utils.input_fn(
+                args.valid if mode == tf.estimator.ModeKeys.EVAL and args.valid else args.train,
+                args.vocab, args.norm, num_channels=args.num_channels,
+                batch_size=params.batch_size if 'batch_size' in params else args.batch_size,
+                num_epochs=args.num_epochs if mode == tf.estimator.ModeKeys.TRAIN else 1,
+                num_parallel_calls=args.num_parallel_calls,
```
Comment (akademi4eg) [inline train.py]: Shouldn't t2t and non-t2t versions handle `num_parallel_calls` in the same way?

File: utils/dataset_utils.py [hunk 5]
```diff
@@ -1,20 +1,134 @@
...
 __all__ = [
-    'read_dataset',
-    'process_dataset',
+    'input_fn',
+    'input_fn_t2t'
 ]
 
+def read_dataset_t2t_format(data_dir, num_parallel_calls, mode, max_frames, max_symbols):
+    problem = SpeechRecognitionProblem()
+    problem.name = 'asr_ipa_precalc'
```
Comment (akademi4eg) [inline utils/dataset_utils.py]: Probably we should support other problems, e.g. librispeech. So it might be reasonable to move problem name to configuration.

File: utils/params_utils.py [hunk 6]
```diff
@@ -77,8 +79,8 @@ def create_hparams(args, target_vocab_size=None, binf_count=None, sos_id=1, eos_
         is_reset = args.reset
     except AttributeError:
         is_reset = False
-    if os.path.exists(hparams_file) and not is_reset:
-        with open(hparams_file, 'r') as f:
+    if tf.gfile.Exists(hparams_file) and not is_reset:
+        with tf.gfile.Open(hparams_file, 'r') as f:
```
Comment (akademi4eg) [inline utils/params_utils.py]: This function is deprecated, let's switch to newer API: `tf.io.gfile.GFile()`

**Review by akademi4eg — CHANGES_REQUESTED** [review]

File: model_helper.py (hunk 2 above)
Comment (EugKar) [reply model_helper.py]: According to the [documentation](https://www.tensorflow.org/api_docs/python/tf/summary/image), "The images are built from tensor which must be 4-D with shape `[batch_size, height, width, channels]`". We should try whether batch sizes in different hooks may differ, if yes, then it's simpler to do as you have suggested.

File: train.py (hunk 3 above)
Comment (EugKar) [reply train.py:5]: Another option is to copy/paste code from `tensor2tensor`, however, there is too much code  beyond  a simple call to `problem.dataset`, and what is now done by just 7 lines of code in `read_dataset_t2t_format`, may turn into several hundreds of lines. I've tried that )

Comment (EugKar) [reply train.py:5]: Of course, we can do it import in `if` blocks...

File: train.py (hunk 4 above)
Comment (EugKar) [reply train.py]: Why not? The default value should naturally correspond to the number of cores.

File: utils/dataset_utils.py (hunk 5 above)
Comment (EugKar) [reply utils/dataset_utils.py]: So far we don't support `librispeech`, because its dataset format is different from ours. Targets in `librispeech` are decoded as UTF-8 bytes, without using any vocabulary. If we  decide to add support to `librispeech` one day, we'll need to do much more changes to the code than just moving problem name to config. BTW, not only `asr_ipa_precalc` problem is supported now, but all problems which use `IPAEncoder`.

File: utils/params_utils.py (hunk 6 above)
Comment (EugKar) [reply utils/params_utils.py]: There is no mention about deprecation in the documentation.

File: model_helper.py (hunk 2 above)
Comment (akademi4eg) [reply model_helper.py]: I tried double expand dims several times and it worked well.

File: train.py (hunk 3 above)
Comment (akademi4eg) [reply train.py:5]: Do you think creating conversion script is a bad idea (from t2t format)? This would have isolated the dependency.

Copying is definitely not an option. :)

File: train.py (hunk 4 above)
Comment (akademi4eg) [reply train.py]: I mean that t2t version sometimes fallsback to 64, while non-t2t version always uses values from args.

File: utils/dataset_utils.py (hunk 5 above)
Comment (akademi4eg) [reply utils/dataset_utils.py]: UTF-8 bytes is not a problem. We could just use 256 vocabulary size.
Hardcoding for `asr_ipa_precalc` problem makes things rather complicated, because this way this (public!) repository depends not only on tensor2tensor, but also on t2t_problems. Datapreparation becomes a very convoluted process.

File: utils/params_utils.py (hunk 6 above)
Comment (akademi4eg) [reply utils/params_utils.py]: TF 1.14.0:
`W0806 16:56:12.094715 139758041270080 deprecation_wrapper.py:119] From ...: The name tf.gfile.Open is deprecated. Please use tf.io.gfile.GFile instead.`

File: train.py (hunk 3 above)
Comment (EugKar) [reply train.py:5]: The only benefit of that is features precalculation at conversion stage, which is very fast at T2T. However, we'll need to decode targets from numbers to characters again )

File: utils/dataset_utils.py (hunk 5 above)
Comment (EugKar) [reply utils/dataset_utils.py]: `problem.name` is only used for fie names :)

File: train.py (hunk 4 above)
Comment (EugKar) [reply train.py]: Inside `tensor2tensor` it is set to 64 if TPU is used.

File: utils/dataset_utils.py (hunk 5 above)
Comment (akademi4eg) [reply utils/dataset_utils.py]: Then it definitely should be a configurable option!

**Review by akademi4eg — APPROVED** [review]
